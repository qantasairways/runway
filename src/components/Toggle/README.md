```js
<Toggle
  label="Uncontrolled"
  id="toggle-1"
  onChange={value => console.log(value)}
  containerClassName="toggle-container-name-1"
/>
```

```js
<StatefulManager initial>
  {({ value, updater }) => (
    <Toggle
      label="Controlled"
      id="toggle-3"
      checked={value}
      onChange={() => updater(!value)}
      containerClassName="toggle-container-name-3"
    />
  )}
</StatefulManager>
```
