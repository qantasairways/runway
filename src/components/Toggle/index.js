import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Switch from 'react-switch';
import { colours, fontSize } from '../../theme/airways';

const SELECTORS = {
  REACT_SWITCH: {
    ELEMENT_CONTAINER: '.react-switch',
    BACKGROUND: '.react-switch-bg'
  }
};

const LocalStylesInjector = ({
  height,
  children,
  checked,
  containerClassName
}) => {
  const thickenBorder = () => {
    const existingTransitions = `background 150ms ease 0s`;
    const offBoxShadow = `inset ${colours.darkGrey} 0px 0px 1px 3px`;
    return {
      [SELECTORS.REACT_SWITCH.BACKGROUND]: {
        transition: `${existingTransitions}, box-shadow 150ms ease 0s !important`,
        boxShadow: checked ? 'none' : offBoxShadow
      }
    };
  };

  const keepAllVertAligned = {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center'
  };

  return (
    <div
      className={containerClassName}
      css={{
        [SELECTORS.REACT_SWITCH.ELEMENT_CONTAINER]: {
          marginLeft: '8px',
          backgroundColor: colours.darkerGrey,
          height: `${height}px !important`
        },
        backgroundColor: colours.darkerGrey,
        ...thickenBorder(),
        ...keepAllVertAligned
      }}
    >
      {children}
    </div>
  );
};

LocalStylesInjector.propTypes = {
  height: PropTypes.number.isRequired,
  children: PropTypes.node.isRequired,
  checked: PropTypes.bool.isRequired,
  containerClassName: PropTypes.string.isRequired
};

const LabelText = ({ children }) => (
  <span
    css={{
      textTransform: 'uppercase',
      color: colours.white,
      fontSize: fontSize.label
    }}
  >
    {children}
  </span>
);

LabelText.propTypes = {
  children: PropTypes.node.isRequired
};

class Toggle extends Component {
  static propTypes = {
    id: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    containerClassName: PropTypes.string,
    onChange: PropTypes.func.isRequired,
    checked: PropTypes.bool.isRequired
  };

  static defaultProps = {
    containerClassName: 'react-toggle-container'
  };

  static isControlled = props => typeof props.checked === 'boolean';

  state = {
    checked: false
  };

  handleChange = () => {
    if (Toggle.isControlled(this.props)) {
      this.props.onChange();
    } else {
      const update = !this.state.checked; // eslint-disable-line
      this.setState({ checked: update }, () => this.props.onChange(update));
    }
  };

  getCheckedState = () =>
    Toggle.isControlled(this.props) ? this.props.checked : this.state.checked;

  render = () => {
    const {
      id,
      label,
      containerClassName = 'react-toggle-container'
    } = this.props;
    return (
      <LocalStylesInjector
        containerClassName={containerClassName}
        height={30}
        checked={this.getCheckedState()}
      >
        {/* eslint-disable-next-line jsx-a11y/label-has-for */}
        <label htmlFor={id}>
          <LabelText>{label}</LabelText>
        </label>
        <Switch
          checked={this.getCheckedState()}
          onChange={this.handleChange}
          handleDiameter={26}
          offColor={colours.darkerGrey}
          onColor={colours.highlights}
          offHandleColor={colours.white}
          onHandleColor={colours.white}
          boxShadow="0px 0px 1px 0px rgba(0, 0, 0, 0.3)"
          uncheckedIcon={false}
          checkedIcon={false}
          height={30}
          width={48}
          className="react-switch"
          id={id}
        />
      </LocalStylesInjector>
    );
  };
}

export default Toggle;
