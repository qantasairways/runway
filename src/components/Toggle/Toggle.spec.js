import React from 'react';
import { shallow } from 'enzyme';
import Toggle from './index';

describe('Toggle', () => {
  describe('Uncontrolled', () => {
    it('should render correctly', () => {
      const toggle = shallow(
        <Toggle
          id="mockId-1"
          onChange={() => {}}
          label="mockLabel"
          containerClassName="mockContainerClassName"
        />
      );
      expect(toggle).toMatchSnapshot();
    });
  });

  describe('Controlled', () => {
    it('should render checked="true" correctly', () => {
      const toggle = shallow(
        <Toggle
          id="mockId-2"
          onChange={() => {}}
          checked
          label="mockLabel"
          containerClassName="mockContainerClassName"
        />
      );
      expect(toggle).toMatchSnapshot();
    });
    it('should render checked="false" correctly', () => {
      const toggle = shallow(
        <Toggle
          id="mockId-3"
          onChange={() => {}}
          checked={false}
          label="mockLabel"
          containerClassName="mockContainerClassName"
        />
      );
      expect(toggle).toMatchSnapshot();
    });
  });
});
