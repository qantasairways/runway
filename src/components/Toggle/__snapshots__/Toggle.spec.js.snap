// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle Controlled should render checked="false" correctly 1`] = `
<LocalStylesInjector
  checked={false}
  containerClassName="mockContainerClassName"
  height={30}
>
  <label
    htmlFor="mockId-3"
  >
    <LabelText>
      mockLabel
    </LabelText>
  </label>
  <ReactSwitch
    activeBoxShadow="0 0 2px 3px #3bf"
    boxShadow="0px 0px 1px 0px rgba(0, 0, 0, 0.3)"
    checked={false}
    checkedIcon={false}
    className="react-switch"
    disabled={false}
    handleDiameter={26}
    height={30}
    id="mockId-3"
    offColor="#323232"
    offHandleColor="#fff"
    onChange={[Function]}
    onColor="#8de2e0"
    onHandleColor="#fff"
    uncheckedIcon={false}
    width={48}
  />
</LocalStylesInjector>
`;

exports[`Toggle Controlled should render checked="true" correctly 1`] = `
<LocalStylesInjector
  checked={true}
  containerClassName="mockContainerClassName"
  height={30}
>
  <label
    htmlFor="mockId-2"
  >
    <LabelText>
      mockLabel
    </LabelText>
  </label>
  <ReactSwitch
    activeBoxShadow="0 0 2px 3px #3bf"
    boxShadow="0px 0px 1px 0px rgba(0, 0, 0, 0.3)"
    checked={true}
    checkedIcon={false}
    className="react-switch"
    disabled={false}
    handleDiameter={26}
    height={30}
    id="mockId-2"
    offColor="#323232"
    offHandleColor="#fff"
    onChange={[Function]}
    onColor="#8de2e0"
    onHandleColor="#fff"
    uncheckedIcon={false}
    width={48}
  />
</LocalStylesInjector>
`;

exports[`Toggle Uncontrolled should render correctly 1`] = `
<LocalStylesInjector
  checked={false}
  containerClassName="mockContainerClassName"
  height={30}
>
  <label
    htmlFor="mockId-1"
  >
    <LabelText>
      mockLabel
    </LabelText>
  </label>
  <ReactSwitch
    activeBoxShadow="0 0 2px 3px #3bf"
    boxShadow="0px 0px 1px 0px rgba(0, 0, 0, 0.3)"
    checked={false}
    checkedIcon={false}
    className="react-switch"
    disabled={false}
    handleDiameter={26}
    height={30}
    id="mockId-1"
    offColor="#323232"
    offHandleColor="#fff"
    onChange={[Function]}
    onColor="#8de2e0"
    onHandleColor="#fff"
    uncheckedIcon={false}
    width={48}
  />
</LocalStylesInjector>
`;
